modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -1317,152 +1379,153 @@ private RuntimeException convertThrowable(Throwable th) {
+                });
+
+                if (plannedEntry.value() != null) {
+                    if (!metaStorageMgr.invoke(If.iif(
+                            revision(partAssignmentsPlannedKey(partId)).eq(plannedEntry.revision()),
+                            ops(
+                                    put(partAssignmentsStableKey(partId), pendingEntry.value()),
+                                    put(partAssignmentsPendingKey(partId), plannedEntry.value()),
+                                    remove(partAssignmentsPlannedKey(partId)))
+                                    .yield(true),
+                            ops().yield(false))).join().getAsBoolean()) {
+                        onNewPeersConfigurationApplied(peers);
+                    }
+                } else {
+                    if (!metaStorageMgr.invoke(If.iif(
+                            notExists(partAssignmentsPlannedKey(partId)),
+                            ops(put(partAssignmentsStableKey(partId), pendingEntry.value()),
+                                remove(partAssignmentsPendingKey(partId))).yield(true),
+                            ops().yield(false))).join().getAsBoolean()) {
+                        onNewPeersConfigurationApplied(peers);
+                    }
+                }
+            }
+
+            @Override
+            public void onReconfigurationError(Status status) {}
+        };
     }
 
     /**
-     * Internal method for setting a baseline.
-     *
-     * @param nodes Names of baseline nodes.
+     * Register the new meta storage listener for changes in pending partitions.
      */
-    private void setBaselineInternal(Set<String> nodes) {
-        if (nodes == null || nodes.isEmpty()) {
-            throw new IgniteException("New baseline can't be null or empty");
-        }
+    private void registerRebalanceListeners() {
+        metaStorageMgr.registerWatchByPrefix(ByteArray.fromString(PENDING_ASSIGNMENTS_PREFIX), new WatchListener() {
+            @Override
+            public boolean onUpdate(@NotNull WatchEvent evt) {
+                assert evt.single();
 
-        var currClusterMembers = new HashSet<>(baselineMgr.nodes());
+                if (evt.entryEvent().newEntry().value() == null) {
+                    return true;
+                }
 
-        var currClusterMemberNames =
-                currClusterMembers.stream().map(ClusterNode::name).collect(Collectors.toSet());
+                int part = extractPartitionNumber(evt.entryEvent().newEntry().key());
+                UUID tblId = extractTableId(evt.entryEvent().newEntry().key());
 
-        for (String nodeName : nodes) {
-            if (!currClusterMemberNames.contains(nodeName)) {
-                throw new IgniteException("Node '" + nodeName + "' not in current network cluster membership. "
-                        + " Adding not alive nodes is not supported yet.");
-            }
-        }
+                TableImpl tbl = tablesByIdVv.latest().get(tblId);
```

Comment:
Why the table should always here?

Reply:
This call states inside the watch for partition pending assignments.

As far as I understand, according to the nature of ordered metastore events:
- this event can't be triggered before table creation (rebalance can't be triggered before it)
- this event can't be triggered after table drop (rebalance for partition can't be triggered for removed table)